From fdef3554192a498e427764dee92cc67bcb66344c Mon Sep 17 00:00:00 2001
From: Chandrakala Chavva <cchavva@caviumnetworks.com>
Date: Fri, 27 May 2016 15:45:03 -0700
Subject: [PATCH 0489/1239] cavium: Fix issue with UART clock

Define CONFIG_OF_BOARD_SETUP to update refclkuaa. Also fix
clock-frequence for sclk by reading RST_BOOT CSR.

Signed-off-by: Chandrakala Chavva <cchavva@caviumnetworks.com>
---
 board/cavium/thunderx/fdt.c     | 17 ++++++++++++++++-
 include/configs/thunderx_81xx.h |  2 +-
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/board/cavium/thunderx/fdt.c b/board/cavium/thunderx/fdt.c
index b6223c52de..534b9c19f5 100644
--- a/board/cavium/thunderx/fdt.c
+++ b/board/cavium/thunderx/fdt.c
@@ -116,6 +116,7 @@ static void ft_setup_uaa_clk(void *fdt)
 	unsigned long baud = CONFIG_BAUDRATE;
 	uint32_t refclk;
 	int nodeoffset, err;
+	uint64_t rst_boot;
 
 	ibrd = CSR_READ_PA(0, UAAX_IBRD(0));
 	ibrd = ibrd & 0xffff;
@@ -125,12 +126,26 @@ static void ft_setup_uaa_clk(void *fdt)
 
 	refclk = (baud * (64 * ibrd + fbrd)) / 4;
 
+	/* Set RCLK */
 	nodeoffset = fdt_path_offset(fdt, "/soc/refclkuaa");
 
 	err = fdt_setprop_u32(fdt, nodeoffset, "clock-frequency", refclk);
 
 	if(err < 0) {
-		printf("WARNING: could not set %s: %s.\n", "clock-frequency",
+		printf("WARNING: could not set %s for RCLK: %s.\n", "clock-frequency",
+				fdt_strerror(err));
+		return;
+	}
+
+	/* Set SCLK (rst_boot.pnr_mul */
+	rst_boot = CSR_READ_PA(0, RST_BOOT);
+	rst_boot = ((rst_boot >> 33) & 0x3f);
+
+	nodeoffset = fdt_path_offset(fdt, "/soc/sclk");
+	err = fdt_setprop_u32(fdt, nodeoffset, "clock-frequency", rst_boot * 50000000ull);
+
+	if(err < 0) {
+		printf("WARNING: could not set %s for SCLK: %s.\n", "clock-frequency",
 				fdt_strerror(err));
 		return;
 	}
diff --git a/include/configs/thunderx_81xx.h b/include/configs/thunderx_81xx.h
index 52673e2218..c880c9a93b 100644
--- a/include/configs/thunderx_81xx.h
+++ b/include/configs/thunderx_81xx.h
@@ -78,7 +78,7 @@
 
 /* Flat Device Tree Definitions */
 #define CONFIG_OF_LIBFDT
-/* #define CONFIG_OF_BOARD_SETUP */
+#define CONFIG_OF_BOARD_SETUP
 
 /* SMP Spin Table Definitions */
 #define CPU_RELEASE_ADDR		(CONFIG_SYS_SDRAM_BASE + 0x7fff0)
-- 
2.29.0

